src/coreclr/jit/gentree.h

```diff
@@ -9613,6 +9613,246 @@ enum varRefKinds
+
+    if (scalar)
+    {
+        count = 1;
+
+#if defined(TARGET_XARCH)
+        // scalar operations on xarch copy the upper bits from arg0
+        *result = arg0;
+#endif // TARGET_XARCH
+    }
+
+    for (uint32_t i = 0; i < count; i++)
+    {
+        // Safely execute `result[i] = oper(arg0[i])`
+
+        TBase input0;
+        memcpy(&input0, &arg0.u8[i * sizeof(TBase)], sizeof(TBase));
+
+        TBase output = EvaluateUnaryScalar<TBase>(oper, input0);
+        memcpy(&result->u8[i * sizeof(TBase)], &output, sizeof(TBase));
+    }
+}
+
+template <typename TSimd>
+void EvaluateUnarySimd(genTreeOps oper, bool scalar, var_types baseType, TSimd* result, TSimd arg0)
+{
+    switch (baseType)
+    {
+        case TYP_FLOAT:
+        {
+            EvaluateUnarySimd<TSimd, float>(oper, scalar, result, arg0);
+            break;
+        }
+
+        case TYP_DOUBLE:
+        {
+            EvaluateUnarySimd<TSimd, double>(oper, scalar, result, arg0);
+            break;
+        }
+
+        case TYP_BYTE:
+        {
+            EvaluateUnarySimd<TSimd, int8_t>(oper, scalar, result, arg0);
+            break;
+        }
+
+        case TYP_SHORT:
+        {
+            EvaluateUnarySimd<TSimd, int16_t>(oper, scalar, result, arg0);
+            break;
+        }
+
+        case TYP_INT:
+        {
+            EvaluateUnarySimd<TSimd, int32_t>(oper, scalar, result, arg0);
+            break;
+        }
+
+        case TYP_LONG:
+        {
+            EvaluateUnarySimd<TSimd, int64_t>(oper, scalar, result, arg0);
+            break;
+        }
+
+        case TYP_UBYTE:
+        {
+            EvaluateUnarySimd<TSimd, uint8_t>(oper, scalar, result, arg0);
+            break;
+        }
+
+        case TYP_USHORT:
+        {
+            EvaluateUnarySimd<TSimd, uint16_t>(oper, scalar, result, arg0);
+            break;
+        }
+
+        case TYP_UINT:
+        {
+            EvaluateUnarySimd<TSimd, uint32_t>(oper, scalar, result, arg0);
+            break;
+        }
+
+        case TYP_ULONG:
+        {
+            EvaluateUnarySimd<TSimd, uint64_t>(oper, scalar, result, arg0);
+            break;
+        }
+
+        default:
+        {
+            unreached();
+        }
+    }
+}
+
+template <typename TBase>
+TBase EvaluateBinaryScalar(genTreeOps oper, TBase arg0, TBase arg1)
+{
+    switch (oper)
+    {
+        case GT_ADD:
+        {
+            return arg0 + arg1;
+        }
+
+        case GT_SUB:
+        {
+            return arg0 - arg1;
+        }
+
+        default:
+        {
+            unreached();
+        }
+    }
+}
+
+template <typename TSimd, typename TBase>
+void EvaluateBinarySimd(genTreeOps oper, bool scalar, TSimd* result, TSimd arg0, TSimd arg1)
+{
+    uint32_t count = sizeof(TSimd) / sizeof(TBase);
+
+    if (scalar)
+    {
+        count = 1;
+
+#if defined(TARGET_XARCH)
+        // scalar operations on xarch copy the upper bits from arg0
+        *result = arg0;
```

Comment:
can you explain me please this part on an example? (the difference between xarch and arm)

Reply:
`xarch` has the behavior where scalar operations "copy the upper bits", that is `x + y` is equivalent to:
```csharp
Vector128<T> result = x;
return result.WithElement(0, x.GetElement(0) + y.GetElement(0));
```

`arm` on the other hand zeros the upper bits, that is `x + y` is equivalent to:
```csharp
Vector128<T> result = Vector128<T>.Zero;
return result.WithElement(0, x.GetElement(0) + y.GetElement(0));
```

Reply:
Added a path that explicitly zeros for Arm64 to help clarify the logic